server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,403 @@
+
+    /**
+     * Returns a set of files currently pending deletion in this directory.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public Set<String> getPendingDeletions() throws IOException {
+        writeLock.lock();
+        try {
+            return localDirectory.getPendingDeletions();
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Function to perform operations once files have been uploaded to Remote Store
+     * Currently deleting the local files here, as once uploaded to Remote, local files become eligible for eviction from FileCache
+     * @param files : recent files which have been successfully uploaded to Remote Store
+     * @throws IOException in case of I/O error
+     */
+    public void afterSyncToRemote(Collection<String> files) throws IOException {
+        logger.trace("afterSyncToRemote called for {}", files);
+        if (remoteDirectory == null) {
+            logger.trace("afterSyncToRemote called even though remote directory is not set");
+            return;
+        }
+        for (String fileName : files) {
+            writeLock.lock();
+            try {
+                /**
+                 * TODO - Unpin the files here from FileCache so that they become eligible for eviction, once pinning/unpinning support is added in FileCache
+                 * Uncomment the below commented line(to remove the file from cache once uploaded) to test block based functionality
+                 */
+                logger.trace("File {} uploaded to Remote Store and now can be eligible for eviction in FileCache", fileName);
+                // fileCache.remove(localDirectory.getDirectory().resolve(fileName));
+            } finally {
+                writeLock.unlock();
+            }
+        }
+    }
+
+    private boolean isTempFile(String name) {
+        return name.endsWith(".tmp");
+    }
+
+    /**
+     * Return the list of files present in Remote
+     */
+    private String[] getRemoteFiles() throws IOException {
+        String[] remoteFiles;
+        try {
+            remoteFiles = remoteDirectory.listAll();
+        } catch (NullPointerException e) {
+            /*
+             * There are two scenarios where the listAll() call on remote directory returns NullPointerException:
+             * - When remote directory is not set
+             * - When init() of remote directory has not yet been called
+             *
+             * Returning an empty list in these scenarios
+             */
+            remoteFiles = new String[0];
+        }
+        return remoteFiles;
+    }
+
+    private void cacheFile(String name) throws IOException {
+        Path filePath = localDirectory.getDirectory().resolve(name);
+        fileCache.put(filePath, new NonBlockCachedIndexInput(localDirectory.openInput(name, IOContext.READ)));
+        // Decrementing ref here as above put call increments the ref of the key
+        fileCache.decRef(filePath);
```

Comment:
Done that as a temp measure in the latest revision. Will add file pinning in File Cache and then pin the file here so that it cannot be evicted unless uploaded to Remote store (where we will unpin it)